docs/types.rst

```diff
@@ -646,10 +646,12 @@ Forced data location:
 Arrays
 ------
 
-Arrays can have a compile-time fixed size or they can be dynamic.
-For storage arrays, the element type can be arbitrary (i.e. also other
-arrays, mappings or structs). For memory arrays, it cannot be a mapping and
-has to be an ABI type if it is an argument of a publicly-visible function.
+Arrays can have a compile-time fixed size or a dynamic size. For arrays stored
+in the ``storage`` data location, the element type is arbitrary (i.e., also
```

Comment:
This reads rather weird if put at the beginning of the section. I would say let's rather talk about arrays in general, how to write them and then - if needed at this point at all - talk about the restrictions. It might even make more sense to talk about ABI types at the point where we talk about external function calls instead at the types section, in the sense of "you cannot use this type in external function calls because it contains e.g. a mapping"

Reply:
@chriseth I agree, but this text was here already and isn't related to the issue this PR closes (I don't think?). I can address it here, or better, create a new PR?

Reply:
Ok, but if you want to argue about just solving the single issue, I would expect adding the link as the only change to this paragraph.

Reply:
It mostly is, plus some clarifying of the text around it based on changes made in other PRs, for consistency etc. It's the same approach taken in other PRs.

Reply:
To be honest, I think it is better to remove the mention of this restriction here. This is the second sentence about arrays and it does not even mention how to write them.

Reply:
Also the restrictions mentioned here are not restrictions of the array type but rather restrictions of the types used inside the arrays. You cannot use these restricted types in isolation, so you cannot use them in arrays either.

Reply:
@chriseth I guess this starts to maybe become the territory of a new PR? How about to close this one, I strip it back to just adding the link, then create a new one with my changes and the suggestion you mention above?

Reply:
Sounds good!